### PR TITLE
Rt 26 move player system

### PR DESCRIPTION
**Pull request by BaptisteLaran**
## Describe your changes

Add a system that handle player movement

## Issue ticket number and link

https://uwuclub.atlassian.net/browse/RT-26?atlOrigin=eyJpIjoiNjA2ZjU1MjdlOGY2NGQwYzk4MGZhZmM3MjRkMzUwZmYiLCJwIjoiaiJ9

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] I have documented my code
- [ ] Do we need to implement unit tests?

### CODE REVIEW

File: src/client/System.cpp [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+#include <SDL2/SDL_image.h>
+#include <unordered_map>
+
+namespace ECS {
+    void System::movePlayer(Core::World &world, Core::SparseArray<Utils::Vector2f> &aPos,
+                            Core::SparseArray<Utils::Speed> &aSpeed, Core::SparseArray<Utils::TypeEntity> &aType)
+    {
+        Event::EventManager *eventManager = Event::EventManager::getInstance();
+        auto keyboardEvent = eventManager->getEventsByType(Event::EventType::KEYBOARD);
+        std::unordered_map<const Event::KeyIdentifier, std::function<void(float &, Utils::Vector2f &)>> keyMap = {
```
Comment (GlassAlo) [inline src/client/System.cpp]: static const

**Review by GlassAlo — DISMISSED** [review]

**Review by GlassAlo — APPROVED** [review]

**Review by GlassAlo — DISMISSED** [review]

**Review by Beafowl-Pull — DISMISSED** [review]
RAS

**Review by Valegox — DISMISSED** [review]

File: src/client/SDLDisplayClass.hpp [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+** EPITECH PROJECT, 2023
+** R-Type
+** File description:
+** SDLDisplayClass
+*/
+
+#ifndef SDLDISPLAYClass_HPP_
+#define SDLDISPLAYClass_HPP_
+
+#define SCREEN_WIDTH  1920
```
Comment (Valegox) [inline src/client/SDLDisplayClass.hpp]: constexpr

File: src/client/SDLDisplayClass.hpp [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+        /**
+         * @brief Construct a new SDLDisplayClass object
+         *
+         */
+        SDLDisplayClass();
+        SDL_Renderer *_renderer;
+        SDL_Window *_window;
+
+    private:
```
Comment (Valegox) [inline src/client/SDLDisplayClass.hpp:51]: remove this

File: src/client/SDLDisplayClass.hpp [hunk 4]
```diff
@@ -0,0 +1,54 @@
+/*
+** EPITECH PROJECT, 2023
+** R-Type
+** File description:
+** SDLDisplayClass
+*/
+
+#ifndef SDLDISPLAYClass_HPP_
```
Comment (Valegox) [inline src/client/SDLDisplayClass.hpp:8]: upper case

File: src/client/System.cpp [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        Event::EventManager *eventManager = Event::EventManager::getInstance();
+        SDL_Event event;
+
+        while (SDL_PollEvent(&event)) {
+            if (event.type == SDL_KEYDOWN && (_keyMap.find(event.key.keysym.sym) != _keyMap.end())) {
+                auto keyEvent = new Event::KeyboardEvent(_keyMap.at(event.key.keysym.sym), Event::KeyState::PRESSED);
+                eventManager->pushEvent(keyEvent);
+            }
+            if (event.type == SDL_WINDOWEVENT && (_windowEventMap.find(event.window.event) != _windowEventMap.end())) {
+                auto windowEvent = new Event::WindowEvent(1920, 1920, 0, 0, ECS::Event::WindowDisplayState::FULLSCREEN,
```
Comment (Valegox) [inline src/client/System.cpp]: 1920x1920 ? should it be 1920x1080 ? and use a constexpr pls

File: src/client/System.cpp [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+#include <SDL_image.h>
+#include <unordered_map>
+
+namespace ECS {
+    void System::movePlayer(Core::World &world, Core::SparseArray<Utils::Vector2f> &aPos,
+                            Core::SparseArray<Utils::Speed> &aSpeed, Core::SparseArray<Utils::TypeEntity> &aType)
+    {
+        Event::EventManager *eventManager = Event::EventManager::getInstance();
+        auto keyboardEvent = eventManager->getEventsByType(Event::EventType::KEYBOARD);
+        static const std::unordered_map<const Event::KeyIdentifier, std::function<void(float &, Utils::Vector2f &)>>
```
Comment (GlassAlo) [inline src/client/System.cpp]: not const key, just const map

File: src/client/System.cpp [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+#include <SDL.h>
+#include <functional>
+#include "EventManager.hpp"
+#include "KeyboardEvent.hpp"
+#include "SDLDisplayClass.hpp"
+#include <SDL_image.h>
+#include <unordered_map>
+
+namespace ECS {
+    void System::movePlayer(Core::World &world, Core::SparseArray<Utils::Vector2f> &aPos,
```
Comment (GlassAlo) [inline src/client/System.cpp]: Own file

File: src/client/System.cpp [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+                continue;
+            }
+            for (auto event : keyboardEvent) {
+                auto keyEvent = static_cast<Event::KeyboardEvent *>(event);
+                keyMap.at(keyEvent->_keyId)(aSpeed[i].value().speed, aPos[i].value());
+            }
+        }
+    }
+
+    void System::getInput(Core::World &aWorld)
```
Comment (GlassAlo) [inline src/client/System.cpp]: Own file (System+GetInput.cpp)

File: src/client/System.cpp [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+            if (event.type == SDL_WINDOWEVENT && (_windowEventMap.find(event.window.event) != _windowEventMap.end())) {
+                auto windowEvent = new Event::WindowEvent(
+                    SCREEN_WIDTH, SCREEN_WIDTH, 0, 0, ECS::Event::WindowDisplayState::FULLSCREEN,
+                    ECS::Event::WindowFocusState::FOCUSED, _windowEventMap.at(event.window.event));
+                eventManager->pushEvent(windowEvent);
+            }
+        }
+    }
+
+    void System::quitSDL(Core::World &aWorld)
```
Comment (GlassAlo) [inline src/client/System.cpp]: Own file (System+QuitSDL.cpp)

File: src/client/System.cpp [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+        for (size_t i = 0; i < windowEvent.size(); i++) {
+            auto event = static_cast<Event::WindowEvent *>(windowEvent[i]);
+            if (event->_windowEventType == Event::WindowEventType::CLOSED) {
+                display.~SDLDisplayClass();
+                break;
+            }
+        }
+    }
+
+    void System::loadTextures(Core::World &aWorld, Core::SparseArray<Utils::LoadedSprite> &aSprites)
```
Comment (GlassAlo) [inline src/client/System.cpp]: Own file (System+LoadTextures.cpp)

File: src/client/System.cpp [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+            if (!aSprites[i].has_value() || aSprites[i].value().texture != nullptr)
+                continue;
+            aSprites[i].value().texture = IMG_LoadTexture(display._renderer, aSprites[i].value().path.c_str());
+            if (aSprites[i].value().texture == nullptr) {
+                std::cerr << "Failed to load texture: " << aSprites[i].value().path << std::endl;
+            }
+        }
+    }
+
+    void System::displayEntities(Core::World &aWorld, Core::SparseArray<Utils::LoadedSprite> &aSprites,
```
Comment (GlassAlo) [inline src/client/System.cpp]: Own file (System+DisplayEntities.cpp)

**Review by GlassAlo — CHANGES_REQUESTED** [review]

**Review by GlassAlo — APPROVED** [review]

**Pull request merged**